File: server/libs/os-services/service-manager.js
const path = require('path');

const _services = Symbol('services');

const Service = require(path.join(__dirname, "data", "service"));
const Monitor = require(path.join(__dirname, "util", "monitor"));

class ServiceManager extends Monitor {
    constructor() {
        super();
        this[_services] = null;
    }

    get Services() { return this[_services]; }

    update() { super.update(); }
}

// If the OS is microsoft windows
if (/^win/.test(process.platform)) {
    const PowerShell = require(path.join(__dirname, "windows", "powershell"));
    const PS = new PowerShell();

    const CMD_ListServices = "Get-WmiObject win32_service | select Name, DisplayName, Description, State, @{Name=\"Path\"; Expression={$_.PathName.split('\"')[1]}}";

    class WindowsServiceManager extends ServiceManager {
        update() {
            var rawServices = PS.execToJSONSync(CMD_ListServices);
            var services = [];

            for (var i in rawServices) {
                var service = new Service(rawServices[i].Name, rawServices[i].Path);
                service.FullName = rawServices[i].DisplayName;
                service.Description = rawServices[i].Description;
                service.State = (rawServices[i].State == "Running" ? Service.States.RUNNING : (rawServices[i].State == "Stopped" ? Service.States.STOPPED : Service.States.UNKNOWN));
                services.push(service);
            }

            this[_services] = services;
            super.update();
        }
    }

    module.exports = WindowsServiceManager;
}
else {
    throw new Error("ServiceManager is not implemented for the operative system " + process.platform);
}

File: server/bin/lib/os-utils/service-manager.dart
import 'dart:io';
import './data/service.dart';
import './util/windows/powershell.dart';
import './util/monitor.dart';

class ServiceManager extends Monitor {
  List<Service> _services = new List<Service>();

  List<Service> get services => this._services;

  static ServiceManager GetOSManager() {
    if (new RegExp("/^windows/").hasMatch(Platform.operatingSystem)) {
      return new WindowsServiceManager();
    } else {
      throw new StateError(
          "ServiceManager is not implemented for the operative system " +
              Platform.operatingSystem);
    }
  }
}

class WindowsServiceManager extends ServiceManager {
  static final String _cmdListServices =
      "Get-WmiObject win32_service | select Name, DisplayName, Description, State, PathName | Format-List";

  void Update() {
    IPowerShellResponse response = PowerShell.execSync(WindowsServiceManager._cmdListServices);
    List<Map<String, dynamic>> rawProcesses = PowerShellResponseParser.fromFormatList(response.toString());
    List<Service> services = new List<Service>();

    for (Map<String, dynamic> rawProcess in rawProcesses) {
      var srv = new Service(rawProcess["Name"], rawProcess["PathName"]);
      srv.FullName = rawProcess.containsKey("DisplayName") ? rawProcess["DisplayName"] : rawProcess["Name"];
      srv.Description = rawProcess.containsKey("Description") ? rawProcess["Description"] : "";
      srv.Status = Service.parseStatus(rawProcess["State"]);
      services.add(srv);
    }

    this._services = services;
    super.Update();
  }
}
